src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs

```diff
@@ -316,80 +318,77 @@ private void BuildFallbackBlock(TagHelperAttributeList attributes, TagHelperCont
                         {
                             addSrc = false;
-                            AppendEncodedVersionedSrc(attribute.Name, src, builder, generateForDocumentWrite: true);
+                            WriteEncodedVersionedSrc(attribute.Name, src, attribute.ValueStyle, StringWriter);
                         }
                     }
 
                     if (addSrc)
                     {
-                        AppendEncodedVersionedSrc(SrcAttributeName, src, builder, generateForDocumentWrite: true);
+                        WriteEncodedVersionedSrc(SrcAttributeName, src, HtmlAttributeValueStyle.DoubleQuotes, StringWriter);
                     }
 
-                    builder.AppendHtml("><\\/script>");
+                    StringWriter.Write("></script>");
                 }
 
+                var stringBuilder = StringWriter.GetStringBuilder();
+                var scriptTags = stringBuilder.ToString();
+                stringBuilder.Clear();
+                var encodedScriptTags = JavaScriptEncoder.Encode(scriptTags);
+                builder.AppendHtml(encodedScriptTags);
+
                 builder.AppendHtml("\"));</script>");
             }
         }
 
-        private string GetAttributeValue(object value)
+        private string GetVersionedSrc(string srcValue)
         {
-            string stringValue;
-            var htmlString = value as HtmlString;
-            if (htmlString != null)
+            if (AppendVersion == true)
             {
-                // Value likely came from an HTML context in the .cshtml file but may still contain double quotes
-                // since attribute could have been enclosed in single quotes.
-                stringValue = htmlString.Value;
-                stringValue = stringValue.Replace("\"", "&quot;");
+                srcValue = _fileVersionProvider.AddFileVersionToPath(srcValue);
             }
-            else
-            {
-                var writer = StringWriter;
-                RazorPage.WriteTo(writer, HtmlEncoder, value);
 
-                // Value is now correctly HTML-encoded but may still contain double quotes since attribute could
-                // have been enclosed in single quotes and portions that were HtmlStrings are not re-encoded.
-                var builder = writer.GetStringBuilder();
-                builder.Replace("\"", "&quot;");
+            return srcValue;
+        }
 
-                stringValue = builder.ToString();
-                builder.Clear();
-            }
+        private void AppendVersionedSrc(
+            string srcName,
+            string srcValue,
+            HtmlAttributeValueStyle structure,
+            IHtmlContentBuilder builder)
+        {
+            srcValue = GetVersionedSrc(srcValue);
 
-            return stringValue;
+            AppendAttribute(srcName, srcValue, structure, builder);
         }
 
-        private void AppendEncodedVersionedSrc(
+        private void WriteEncodedVersionedSrc(
```

Comment:
`WriteVersionedSrc()` since it (like `AppendVersionedSrc()` no longer does additional encoding work.